File: scripts/gantry/orthogonality/code/affine_leakage.m
function affine_leakage
% AFFINE_LEAKAGE  MATLAB companion to affine_leakage.py (Rule 2, engine cross-check).
%
% SCOPE NOTE. Rule 2 clause 1 requires a dual proof for structural claims ABOUT THE FP
% MODEL, where MATLAB reads kamtin-fp-model/ as an independent source. This derivation
% contains no FP-model transcription: it is abstract algebra on a generic affine
% augmentation. So this file is an ENGINE cross-check (a wrong solve branch, a bad
% simplify), not the two-independent-sources check. Claimed at that strength only.
%
% INCIDENT, 2026-09-04, and the reason for the null-space parameterisation below.
% The first version of this file used solve() for both constraints. For a 4-equation,
% 8-unknown system it reported "solved 8 of 8 entries", i.e. it returned the TRIVIAL
% branch L = M = bf = 0, and likewise K = 0 in Part C. That made the K = 0 control
% vanish for the wrong reason and put MATLAB in disagreement with sympy on two rows.
% Both engines now parameterise each constraint by an explicit null-space basis, which
% is unique up to a change of basis and has no branch to pick. Recorded here because
% Rule 2 clause 6 requires engine failures to be written down rather than argued away.
%
% Mirrors the four parts of the Python:
%   A  pointwise orthogonality, imposed as an identity on the xa = 0 slice, leaves the
%      state-to-physical map K entirely free, and the windowed projection stays nonzero.
%      Control at K = 0 separates propagation (inherited from Gyorok) from the state path.
%   B  the missing term in closed form, verified against the symbolic rollout.
%   C  the extension  Psi_T' A^a K = 0  annihilates it identically.
%   D  the rank / Cayley-Hamilton count that decides penalty vs by-construction.
%
% Run:
%   matlab.exe -batch "cd('scripts/gantry/orthogonality/code'); affine_leakage"

NX = 2; NA = 1; NU = 1; NTH = 1;    % minimal, hand-checkable
T_LIST = [2 3 4];

A  = sym('A',  [NX NX],  'real');
G  = sym('G',  [NA NA],  'real');
F  = sym('F',  [NA NX],  'real');
Hm = sym('Hm', [NA NU],  'real');
bg = sym('bg', [NA 1 ],  'real');
K  = sym('K',  [NX NA],  'real');
L  = sym('L',  [NX NX],  'real');
M  = sym('M',  [NX NU],  'real');
bf = sym('bf', [NX 1 ],  'real');
Ph = sym('Ph', [NX NTH], 'real');   % baseline parameter sensitivity

S = struct('A',A,'G',G,'F',F,'Hm',Hm,'bg',bg,'K',K,'L',L,'M',M,'bf',bf);
bar = repmat('=',1,78);

%% ---------------------------------------------------------------- PART A
fprintf('\n%s\nPART A  NEGATIVE: pointwise does not bound the rollout contribution\n%s\n', bar, bar);

cons = [reshape(Ph.'*L, [],1); reshape(Ph.'*M, [],1); reshape(Ph.'*bf, [],1)];
fprintf('Pointwise constraint as an IDENTITY in (x,u) on the whole slice xa = 0:\n');
fprintf('    Ph.T ( L x + M u + bf ) = 0   for ALL x, u\n');
for r = 1:numel(cons)
    fprintf('        %s = 0\n', char(cons(r)));
end
assert(~any(has(cons, K(:))), 'K appears in the pointwise constraint -- setup is wrong');
fprintf('    K appears in NONE of them: the state-to-physical map is unconstrained.\n');

Nb = null(Ph.');                     % see the INCIDENT note above: no solve() here
nk = size(Nb,2);
Lf  = sym('lf',  [nk NX], 'real');
Mf  = sym('mf',  [nk NU], 'real');
bff = sym('bff', [nk 1 ], 'real');
Ls  = Nb*Lf;   Ms = Nb*Mf;   bfs = Nb*bff;
pwK = [L(:); M(:); bf(:)];
pwV = [Ls(:); Ms(:); bfs(:)];
assert(~nz(expand(Ph.'*Ls)) && ~nz(expand(Ph.'*Ms)) && ~nz(expand(Ph.'*bfs)), ...
       'null-space parameterisation does not satisfy the pointwise constraint');
fprintf('    ker(Ph.T) has dimension %d in R^%d; L, M, bf are parameterised by it, so\n', nk, NX);
fprintf('    the pointwise constraint holds by construction, not by a solve branch.\n');
fprintf('    verified: Ph.T L = Ph.T M = Ph.T bf = 0 identically after substitution.\n');
fprintf('    all %d entries of K remain free.\n', numel(K));

for t = 1:numel(T_LIST)
    T = T_LIST(t);
    [xs, us] = traj(T, NX, NU);
    dx   = rollout(S, T, xs, us);
    Psi  = psiT(A, Ph, T);
    proj = expand(subs(Psi.'*dx, pwK, pwV));
    prop = expand(subs(proj, K(:), zeros(numel(K),1)));   % control: ANN cannot read xa
    st   = expand(proj - prop);
    fprintf('\n  T = %d:  Psi_T.T dx_T with the pointwise constraint enforced exactly\n', T);
    fprintf('      nonzero overall                     : %d\n', nz(proj));
    fprintf('      CONTROL K = 0 (static, Gyorok)      : %d   <- propagation, NOT our gap\n', nz(prop));
    fprintf('      STATE PATH terms carrying K         : %d   <- our gap\n', nz(st));
end

%% ---------------------------------------------------------------- PART B
fprintf('\n%s\nPART B  THE MISSING TERM in closed form\n%s\n', bar, bar);
fprintf('    dx_T^state = sum_{i=0}^{T-2} Lam_T(i) g_i ,  g_i = F x_i + Hm u_i + bg\n');
fprintf('    Lam_T(i)   = sum_{j=i+1}^{T-1} A^(T-1-j) K G^(j-1-i)\n\n');

for t = 1:numel(T_LIST)
    T = T_LIST(t);
    [xs, us] = traj(T, NX, NU);
    dx   = rollout(S, T, xs, us);
    Psi  = psiT(A, Ph, T);
    proj = expand(subs(Psi.'*dx, pwK, pwV));
    st   = expand(proj - subs(proj, K(:), zeros(numel(K),1)));
    closed = sym(zeros(NX,1));
    for i = 0:T-2
        closed = closed + lamT(A,K,G,T,i) * (F*xs{i+1} + Hm*us{i+1} + bg);
    end
    closed = subs(expand(Psi.'*closed), pwK, pwV);
    ok = ~nz(closed - st);
    fprintf('  T = %d:  closed form reproduces the symbolic rollout exactly : %d\n', T, ok);
    assert(ok, 'closed form mismatch at T=%d', T);
end

%% ---------------------------------------------------------------- PART C
fprintf('\n%s\nPART C  THE EXTENSION: orthogonality on the state-to-physical map K\n%s\n', bar, bar);
fprintf('    Psi_T.T A^a K = 0 ,  a = 0 ... T-2                              (EXT)\n');
fprintf('    <=> range(K) orthogonal to span{ Psi_T, A.T Psi_T, ... }\n');
fprintf('    At n_a = 0 there is no K, (EXT) is vacuous, method collapses to Gyorok.\n\n');

for t = 1:numel(T_LIST)
    T   = T_LIST(t);
    Psi = psiT(A, Ph, T);
    Sk  = sym([]);
    for a = 0:T-2
        Sk = [Sk; Psi.'*A^a]; %#ok<AGROW>
    end
    Nk = null(Sk);
    dk = size(Nk,2);
    if dk == 0
        extV = sym(zeros(numel(K),1));   % only K = 0 satisfies (EXT)
        triv = true;
    else
        Kf   = sym('kf', [dk NA], 'real');
        Kp   = Nk*Kf;
        extV = Kp(:);
        triv = ~nz(Kp);
    end

    [xs, us] = traj(T, NX, NU);
    dx   = rollout(S, T, xs, us);
    proj = expand(subs(Psi.'*dx, pwK, pwV));
    st   = expand(proj - subs(proj, K(:), zeros(numel(K),1)));
    killed = ~nz(subs(st, K(:), extV));

    fprintf('  T = %d:  %d scalar conditions on %d entries of K, dim ker(S_T) = %d\n', ...
            T, size(Sk,1)*NA, numel(K), dk);
    fprintf('          state-path term annihilated identically : %d\n', killed);
    fprintf('          forced K = 0 (condition vacuous?)       : %d\n', triv);
    assert(killed, 'EXT fails to annihilate the state path at T=%d', T);
end

%% ---------------------------------------------------------------- PART D
fprintf('\n%s\nPART D  BY-CONSTRUCTION OR PENALTY? the counting that decides it\n%s\n', bar, bar);
fprintf('(EXT): range(K) in ker(S_T), S_T the Krylov stack of Psi_T under A.T\n');
fprintf('Cayley-Hamilton: A.T has degree n_x, so the span saturates at a = n_x - 1.\n\n');
for T = 2:NX+3
    Psi = psiT(A, Ph, T);
    St  = sym([]);
    for a = 0:T-2
        St = [St; Psi.'*A^a]; %#ok<AGROW>
    end
    r = rank(St);
    fprintf('    T = %d:  stack %dx%d   rank = %d   dim ker = %d\n', ...
            T, size(St,1), size(St,2), r, NX - r);
end
fprintf('\n  Saturation is at rank = n_x, so dim ker(S_T) = 0 and (EXT) forces K = 0.\n');
fprintf('  A by-construction K = (I - Pi) Ktilde would delete the additional-state\n');
fprintf('  pathway entirely. The extension must be a SOFT PENALTY over realised data.\n');

fprintf('\nMATLAB run complete: every assertion passed.\n');
end

% ---------------------------------------------------------------- helpers
function b = nz(e)
b = ~isequal(simplify(expand(e)), sym(zeros(size(e))));
end

function [xs, us] = traj(T, NX, NU)
xs = cell(T,1); us = cell(T,1);
for k = 1:T
    xs{k} = sym(sprintf('x%d_', k-1), [NX 1], 'real');
    us{k} = sym(sprintf('u%d_', k-1), [NU 1], 'real');
end
end

function dx = rollout(S, T, xs, us)
% Cumulative ANN contribution to the physical state after T steps, from xa_0 = 0.
xa = sym(zeros(size(S.G,1),1));
f  = cell(T,1);
for k = 1:T
    f{k} = expand(S.K*xa + S.L*xs{k} + S.M*us{k} + S.bf);
    xa   = expand(S.G*xa + S.F*xs{k} + S.Hm*us{k} + S.bg);
end
dx = sym(zeros(size(S.A,1),1));
for j = 1:T
    dx = dx + S.A^(T-j) * f{j};
end
dx = expand(dx);
end

function P = psiT(A, Ph, T)
P = sym(zeros(size(Ph)));
for j = 0:T-1
    P = P + A^(T-1-j) * Ph;
end
P = expand(P);
end

function Lm = lamT(A, K, G, T, i)
Lm = sym(zeros(size(K)));
for j = i+1:T-1
    Lm = Lm + A^(T-1-j) * K * G^(j-1-i);
end
Lm = expand(Lm);
end

File: scripts/gantry/orthogonality/code/affine_leakage.py
"""Extending orthogonal-projection regularisation to an augmentation WITH ADDITIONAL STATES.

Gyorok's penalty constrains the ANN OUTPUT at a point: `Q^T f_ANN(x, u) = 0`. His augmentation
is a static map, so that is the whole of the ANN's effect on the physical states. Ours is not:
the ANN carries its own state `xa`, and the path

        xa  ->  (next step)  ->  physical correction  ->  physical state

is invisible to a penalty evaluated at `xa = 0`. This script does three things, in order:

  PART A  NEGATIVE (motivation, not the contribution).  With the pointwise constraint enforced
          EXACTLY, the augmentation still contributes inside the parameter subspace over a
          rollout. A control separates the two mechanisms:
             K = 0   ANN does not read `xa`  -> propagation only; applies to Gyorok too;
                     NOT our gap
             K != 0  ANN reads `xa`          -> the state path; OUR gap
  PART B  THE MISSING TERM in closed form. Not "leakage exists" but what it equals, so the
          extension has a defined target.
  PART C  THE EXTENSION.  A condition on the state-to-physical map `K`, verified symbolically
          to annihilate the missing term identically, and verified to reduce to Gyorok when
          there are no additional states.
  PART D  BY-CONSTRUCTION OR PENALTY?  Dimension counting on the extension's condition, which
          decides which of the two enforcement routes is even feasible.

MODEL (affine, i.e. the augmentation linearised about the operating point; affine is chosen
because an existence claim proved for the SIMPLEST network transfers to every richer one)

    x_{k+1}  = A x_k  + B u_k + f_k              physical,   n_x
    xa_{k+1} = G xa_k + F x_k + H u_k + bg       additional, n_a          <- Gyorok has none
    f_k      = K xa_k + L x_k + M u_k + bf       correction into x

    K   maps the additional states into the physical correction. At `xa = 0` it is multiplied
        by zero, so the deployed penalty cannot see it. This is the loophole written as a
        parameter block rather than as a claim about hidden-layer weights.

Two choices that make PART A stronger than the deployed situation:
  * the pointwise constraint is imposed as an IDENTITY IN (x, u) over the ENTIRE `xa = 0`
    slice, not at a finite point set, so "sample the slice better" cannot rescue it;
  * the trajectory symbols are free, so every identity holds for ALL trajectories.

Outputs -> affine_leakage.json   (same directory)
Run: conda run -n GraduationProject python scripts/gantry/orthogonality/code/affine_leakage.py
"""
import os
import json
import itertools

import sympy as sp

HERE = os.path.dirname(os.path.abspath(__file__))
OUT = os.path.join(HERE, 'affine_leakage.json')

NX, NA, NU, NTH = 2, 1, 1, 1   # minimal, so every expression stays hand-checkable
T_LIST = [2, 3, 4]

RES = {'dims': dict(nx=NX, na=NA, nu=NU, n_theta=NTH), 'T': T_LIST}


def sym_mat(name, r, c):
    return sp.Matrix(r, c, lambda i, j: sp.Symbol(f'{name}{i}{j}', real=True))


def hdr(s):
    print('\n' + '=' * 78)
    print(s)
    print('=' * 78)


# --------------------------------------------------------------------------- model
A = sym_mat('A', NX, NX)
G = sym_mat('G', NA, NA)
F = sym_mat('F', NA, NX)
H = sym_mat('H', NA, NU)
bg = sym_mat('bg', NA, 1)
K = sym_mat('K', NX, NA)
L = sym_mat('L', NX, NX)
M = sym_mat('M', NX, NU)
bf = sym_mat('bf', NX, 1)
Phi = sym_mat('P', NX, NTH)      # baseline parameter sensitivity, d f_base / d theta


def rollout(T, xs, us, Kmat=K):
    """Cumulative ANN contribution to the physical state after T steps, from xa_0 = 0."""
    xa = sp.zeros(NA, 1)
    f_list = []
    for k in range(T):
        f_list.append(sp.expand(Kmat * xa + L * xs[k] + M * us[k] + bf))
        xa = sp.expand(G * xa + F * xs[k] + H * us[k] + bg)
    dx = sp.zeros(NX, 1)
    for j in range(T):
        dx += A ** (T - 1 - j) * f_list[j]
    return sp.expand(dx)


def psi(T):
    """Windowed parameter sensitivity: Phi propagated over the same window.
    The deployed pointwise Phi is the T = 1 case."""
    S = sp.zeros(NX, NTH)
    for j in range(T):
        S += A ** (T - 1 - j) * Phi
    return sp.expand(S)


# ======================================================================== PART A
hdr('PART A  NEGATIVE: the pointwise constraint does not bound the rollout contribution')

print('Pointwise constraint, imposed as an IDENTITY in (x, u) on the whole slice xa = 0:')
print('    Phi^T ( L x + M u + bf ) = 0     for ALL x, u')
cons = ([sp.Eq(e, 0) for e in Phi.T * L] +
        [sp.Eq(e, 0) for e in Phi.T * M] +
        [sp.Eq(e, 0) for e in Phi.T * bf])
for c in cons:
    print('       ', c)
print('\n    K does not appear in a single one of these equations.')
print('    -> the state-to-physical map is UNCONSTRAINED by the pointwise penalty.')
RES['pointwise_constraints'] = [str(c) for c in cons]

# NOT `sp.solve`. A linear solve may return the TRIVIAL branch (L = M = bf = 0), which
# makes the K = 0 control vanish for the wrong reason; the MATLAB run of 2026-09-04 did
# exactly that and the two engines disagreed until both were moved to this form.
# Instead parameterise the constraint by an explicit null-space basis, which is unique up
# to a change of basis and has no branch to pick: every column of L, M, bf must lie in
# ker(Phi^T), so write them as N * (free).
N = sp.Matrix.hstack(*sp.Matrix(Phi.T).nullspace())
print(f'    ker(Phi^T) has dimension {N.shape[1]} in R^{NX}; L, M, bf are parameterised')
print('    by it, so the pointwise constraint holds by construction, not by a solve branch.')
Lf, Mf, bff = sym_mat('lf', N.shape[1], NX), sym_mat('mf', N.shape[1], NU), sym_mat('bff', N.shape[1], 1)
PW = {}
for src, tgt in ((N * Lf, L), (N * Mf, M), (N * bff, bf)):
    for e, t in zip(src, tgt):
        PW[t] = e
assert sp.expand(Phi.T * L.subs(PW)) == sp.zeros(NTH, NX)
assert sp.expand(Phi.T * M.subs(PW)) == sp.zeros(NTH, NU)
assert sp.expand(Phi.T * bf.subs(PW)) == sp.zeros(NTH, 1)
print('    verified: Phi^T L = Phi^T M = Phi^T bf = 0 identically after substitution.')
print(f'    all {NX * NA} entries of K remain free.')

for T in T_LIST:
    xs = [sym_mat(f'x{k}_', NX, 1) for k in range(T)]
    us = [sym_mat(f'u{k}_', NU, 1) for k in range(T)]
    proj = sp.expand((psi(T).T * rollout(T, xs, us))[0, 0].subs(PW))
    prop = sp.expand(proj.subs({k: 0 for k in K}))    # control: ANN does not read xa
    state = sp.expand(proj - prop)                    # everything carrying a K entry
    print(f'\n  T = {T}:  Psi_T^T dx_T  with the pointwise constraint enforced exactly')
    print(f'      nonzero overall                       : {proj != 0}')
    print(f'      CONTROL  K = 0 (static ANN, Gyorok)   : {prop != 0}'
          '   <- propagation; NOT our gap')
    print(f'      STATE PATH  terms carrying K          : {state != 0}'
          '   <- our gap')
    RES[f'A_T{T}'] = dict(nonzero=bool(proj != 0), control_K0=bool(prop != 0),
                          state_path=bool(state != 0))

print('\n  READ: the control is nonzero too, so propagation alone already breaks the')
print('  pointwise guarantee. That part is inherited from Gyorok and is not what we claim.')
print('  The state-path row is the part that exists ONLY because of the additional states,')
print('  and it is the part the rest of this script removes.')


# ======================================================================== PART B
hdr('PART B  THE MISSING TERM in closed form')

print('xa_j = sum_{i<j} G^(j-1-i) g_i  with  g_i = F x_i + H u_i + bg, from xa_0 = 0.')
print('Substituting into dx_T = sum_j A^(T-1-j) f_j and collecting by g_i:')
print()
print('    dx_T^state = sum_{i=0}^{T-2} Lam_T(i) g_i ,')
print('    Lam_T(i)   = sum_{j=i+1}^{T-1} A^(T-1-j) K G^(j-1-i)')
print()


def lam(T, i):
    S = sp.zeros(NX, NA)
    for j in range(i + 1, T):
        S += A ** (T - 1 - j) * K * G ** (j - 1 - i)
    return sp.expand(S)


for T in T_LIST:
    xs = [sym_mat(f'x{k}_', NX, 1) for k in range(T)]
    us = [sym_mat(f'u{k}_', NU, 1) for k in range(T)]
    proj = sp.expand((psi(T).T * rollout(T, xs, us))[0, 0].subs(PW))
    state = sp.expand(proj - proj.subs({k: 0 for k in K}))
    closed = sp.zeros(NX, 1)
    for i in range(T - 1):
        closed += lam(T, i) * sp.expand(F * xs[i] + H * us[i] + bg)
    ok = sp.simplify(sp.expand((psi(T).T * sp.expand(closed))[0, 0]) - state) == 0
    print(f'  T = {T}:  closed form reproduces the symbolic rollout exactly : {ok}')
    assert ok, f'closed form mismatch at T={T}'
    RES[f'B_T{T}_closed_form_verified'] = bool(ok)

print('\n  Every term is  A^a K G^b  with a + b = T-2-i. So the missing term is set by K (the')
print('  unconstrained block), by how hard the additional state is driven (F, H, bg), and by')
print('  MIXED POWERS of the baseline dynamics A and the additional-state dynamics G. It')
print('  vanishes identically iff K = 0, i.e. iff the augmentation is effectively static.')
RES['leakage_operator'] = 'Lam_T(i) = sum_{j=i+1}^{T-1} A^(T-1-j) K G^(j-1-i)'


# ======================================================================== PART C
hdr('PART C  THE EXTENSION: an orthogonality condition on the state-to-physical map K')

print('Requiring  Psi_T^T dx_T^state = 0  for ALL trajectories forces it term by term.')
print('A sufficient condition, independent of the data and of G:')
print()
print('    Psi_T^T A^a K = 0 ,   a = 0 ... T-2                                    (EXT)')
print('    <=>  range(K)  orthogonal to  span{ Psi_T, A^T Psi_T, ..., (A^(T-2))^T Psi_T }')
print()
print('Gyorok projects the ANN OUTPUT off the parameter subspace. With additional states one')
print('must ALSO project the STATE-TO-PHYSICAL MAP off the PROPAGATED parameter subspace.')
print('At n_a = 0 there is no K, (EXT) is vacuous, and the method collapses to Gyorok.')
print()

for T in T_LIST:
    rows = [psi(T).T * A ** a for a in range(T - 1)]
    S_T = sp.Matrix.vstack(*rows)
    ker = sp.Matrix(S_T).nullspace()
    n_cond = S_T.shape[0] * NA

    xs = [sym_mat(f'x{k}_', NX, 1) for k in range(T)]
    us = [sym_mat(f'u{k}_', NU, 1) for k in range(T)]
    proj = sp.expand((psi(T).T * rollout(T, xs, us))[0, 0].subs(PW))
    state = sp.expand(proj - proj.subs({k: 0 for k in K}))

    if not ker:
        # No nonzero K satisfies (EXT): the condition is satisfiable only by K = 0.
        EXT = {k: 0 for k in K}
        trivial = True
    else:
        Nk = sp.Matrix.hstack(*ker)
        Kf = sym_mat('kf', Nk.shape[1], NA)
        EXT = {t: e for e, t in zip(Nk * Kf, K)}
        trivial = sp.simplify(K.subs(EXT)) == sp.zeros(NX, NA)

    killed = sp.simplify(state.subs(EXT)) == 0
    print(f'  T = {T}:  {n_cond} scalar conditions on {NX * NA} entries of K, '
          f'dim ker(S_T) = {len(ker)}')
    print(f'          state-path term is annihilated identically : {killed}')
    print(f'          forced K = 0 (condition is vacuous?)       : {trivial}')
    assert killed, f'(EXT) fails to annihilate the state path at T={T}'
    RES[f'C_T{T}'] = dict(n_conditions=int(n_cond), dim_ker=len(ker),
                          annihilates=bool(killed), forces_K_zero=bool(trivial))


# ======================================================================== PART D
hdr('PART D  BY-CONSTRUCTION OR PENALTY? the counting that decides it')

print('(EXT) says range(K) lies in the kernel of the stacked map')
print('    S_T = [ Psi_T , A^T Psi_T , ... , (A^(T-2))^T Psi_T ]^T ,')
print('a Krylov stack of Psi_T under A^T. Two facts decide the enforcement route.\n')

print('  (1) SATURATION. By Cayley-Hamilton, A^T has degree n_x, so the span stops growing')
print('      at a = n_x - 1. Beyond T = n_x + 1 the horizon adds NO new conditions.')
print('      Verified below on the symbolic A by rank of the stack:')
for T in range(2, NX + 4):
    rows = []
    for a in range(T - 1):
        rows.append((psi(T).T * A ** a))
    S_T = sp.Matrix.vstack(*rows)
    r = S_T.rank()
    print(f'        T = {T}:  stack {S_T.shape[0]}x{S_T.shape[1]}   rank = {r}'
          f'   dim ker = {NX - r}')
    RES[f'D_T{T}_rank'] = int(r)

print()
print('  (2) THE BINDING COUNT. rank(S_T) <= n_x, and it reaches n_x as soon as the')
print('      propagated sensitivities span the physical state space. When they do,')
print('      dim ker(S_T) = 0 and (EXT) forces K = 0 EXACTLY: a by-construction')
print('      parameterisation K = (I - Pi) K~ would delete the additional-state pathway')
print('      altogether, which is the pathway the ablation shows is worth ~30 percent.')
print()
print('      For the gantry: n_x = 6 routed physical states, Phi of rank 10 in the')
print('      identifiable parameters, so span(Psi) = R^6 already at a = 0.')
print('      -> STRICT BY-CONSTRUCTION ON K IS NOT AVAILABLE TO US.')
print()
print('  CONSEQUENCE. The extension must be enforced the way Gyorok enforces his: as a')
print('  SOFT PENALTY over the realised data, not by construction. The deployed form is')
print()
print('      beta_a * sum_over_windows || Q_Psi^T sum_i Lam_T(i) g_i ||^2')
print()
print('  which is finitely many conditions over the data rather than an identity in the')
print('  state space, so it admits nonzero K. This is an independent argument for the')
print('  penalty route over the orthogonal-by-construction route, arising from the')
print('  additional states specifically and NOT present in the static case.')
RES['D_conclusion'] = ('strict by-construction on K forces K=0 when span(Psi)=R^nx; '
                       'extension must be a soft penalty over realised data')


hdr('SUMMARY OF CLAIMS')
print('  A  pointwise penalty does not bound the rollout contribution        PROVED (symbolic)')
print('  A  a state-path component exists that is absent in the static case  PROVED (control)')
print('  B  that component equals sum_i Lam_T(i) g_i, Lam as above           PROVED (identity)')
print('  C  (EXT) annihilates it identically and reduces to Gyorok at n_a=0  PROVED (symbolic)')
print('  D  (EXT) by construction forces K=0 once span(Psi)=R^nx             PROVED (rank)')
print('  D  therefore the extension is a penalty, not a construction         FOLLOWS from D')
print()
print('  Rule 1: every quantity above is A, Psi (baseline Jacobian, already built in')
print('  gantry_dynamic/orth_penalty.py) or a network parameter. No truth model is used,')
print('  so all of it transfers to Telica.')

with open(OUT, 'w', encoding='utf-8') as fh:
    json.dump(RES, fh, indent=2)
print(f'\nwrote {OUT}')
